docs/design/sum_types.md

```diff
@@ -0,0 +1,231 @@
+the `choice` keyword followed by the name of the type, and then a list of
+alternatives inside curly braces. An alternative declaration consists of the
+alternative name, optionally followed by a parameter list in parentheses. If
+present, the parameter list has the same syntax as in a
+[function declaration](README.md#functions). For example:
+
+```carbon
+choice OptionalI32 {
+  Some(value: i32),
+  None
+}
+```
+
+This declares a sum type named `OptionalI32` with two alternatives: `Some`,
+which holds a single `i32` value (the parameter name `value` has no effect other
+than documentation), and `None`, which is empty. Choice types can also be
+parameterized, [like class types](generics/details.md#parameterized-types):
+
+```carbon
+choice Optional(T:! Type) {
+  Some(value: T),
+  None
+}
+```
+
+A value of a function-like alternative is specified by "calling" it like a
+function, and a value of an empty alternative like `None` is specified by naming
+it:
+
+```carbon
+var my_opt: Optional(i32) = Optional(i32).None;
+my_opt = Optional(i32).Some(42);
+```
+
+The value of a choice type can be inspected using a `match` statement:
+
+```carbon
+match (my_opt) {
+  case .Some(the_value: i32) => {
+    Print(the_value);
+  }
+  case .None => {
+    Print("None");
+  }
+}
+```
+
+## User-defined sum types
+
+`choice` declarations are a convenience shorthand for common use cases, but they
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using an
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants, and the internal storage
+layout will presumably involve untagged unions or some other low-level storage
+primitive which hasn't been designed yet, but the key to defining a sum type's
+interface is enabling it to support pattern matching. To do that, the sum type
+has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+It does so by implementing the `Match` interface, which is defined as follows:
+
+```carbon
+interface Match {
+  let template Continuation:! Type;
+  fn Op[me: Self, C:! Continuation](continuation: Continuation*)
+    -> Continuation.ReturnType;
```

Comment:
I think this should be:
```suggestion
    -> C.ReturnType;
```
... but it strikes me that using the literal identifier `ReturnType` here means that the name `ReturnType` can never be the name of an alternative, which seems unfortunate. I could imagine that coming up in practice in some not-completely-contrived situation:
```
choice PartOfFunctionDeclaration {
  Name(name: String),
  Parameters(params: Pattern),
  ReturnType(type: Expression),
}
```
... but as a matter of principle I don't like having reserved identifiers that are not keywords.

I think we can address this name collision issue. Suppose we define an interface `MatchContinuation` in the standard library:
```
interface MatchContinuation {
  let ReturnType:! Type;
}
```
then we can use this here:
```suggestion
    -> C.(MatchContinuation.ReturnType);
```

A `Match` impl would then use or `extends MatchContinuation;` (if it doesn't have a colliding alternative) or  `impl as MatchContinuation;` (if it does) instead of `let ReturnType:! Type;`:

```
interface Continuation {
  extends MatchContinuation;
  // I'm not sure if the explicit `MatchContinuation.` qualifier here is necessary.
  // Does `extends` cause name lookup in this scope to look in `MatchContinuation`?
  fn Some[addr me: Self*](value: T) -> MatchContinuation.ReturnType;
  fn None[addr me: Self*]() -> MatchContinuation.ReturnType;
}
```